src/shmem_atomic.h

```diff
@@ -74,6 +74,76 @@ shmem_spinlock_fini(shmem_spinlock_t *lock)
     shmem_internal_assertp(lock->enter == lock->exit);
 }
 
+
+#if (defined(__STDC_NO_ATOMICS__) || !defined(HAVE_STDATOMIC_H))
+
+static inline
+void
+shmem_internal_membar(void) {
+#if USE_XPMEM || shmem_internal_thread_level != SHMEM_THREAD_SINGLE 
```

Comment:
You can't combine a preprocessor conditional and a runtime check like this.  The second part will always evaluate to false at compile time since the preprocessor doesn't have a symbol named shmem_internal_thread_level.

Reply:
I think you can get what you're after by defining a top-level macros like this:
```
#if defined(USE_XPMEM)
#define SHMEM_INTERNAL_NEED_MEMBAR 1
#elif defined (ENABLE_THREADS)
#define SHMEM_INTERNAL_NEED_MEMBAR (shmem_internal_thread_level != SHMEM_THREAD_SINGLE)
#else
#define SHMEM_INTERNAL_NEED_MEMBAR 0
#endif
```

Then in each membar function:
```
  if (SHMEM_INTERNAL_NEED_MEMBAR) __sync_synchronize();
```

Reply:
Thanks Jim. To use ```SHMEM_THREAD_SINGLE``` in the macro, we have to include shmem.h header as follows, right?
```
#define SHMEM_INTERNAL_INCLUDE
#include "shmem.h"
```